Add support for setting tabColor on the command line

Don-Vito opened the pull request:
* Add a tabColor parameter to the `new-tab` and `split-panes` command
* Add --tabColor to the command line, to allow bootstrapping with tabs
  of different colors

Add another field to NewTerminalArgs. Use this field to set
StartingTabColor in Terminal. This color gets overridden by the color
defined by the profile / VT, however can be overridden with the color
picker.

Since the color is the property of the Terminal, when defined for the
tab this color is associated only with the first pane/terminal of the
tab. Additional panes will not inherit this color (to prevent advanced
resolution, where we need to resolve between the inherited color and the
one specified for the pane).

## Validation Steps Performed
* UT for parameters parsing
* Running system with several tabs of different colors.
* Adding custom actions with colors
* Performing operations like split pane, duplicate and so on

Closes #8075

Don-Vito commented:
@zadjii-msft - your CI is impossible 😄 All tests past locally but are flaky as hell when running in pipelines.

zadjii-msft commented on src/cascadia/TerminalSettingsModel/ActionArgs.h:
```diff
@@ -89,6 +92,12 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
             JsonUtils::GetValueForKey(json, ProfileIndexKey, args->_ProfileIndex);
             JsonUtils::GetValueForKey(json, ProfileKey, args->_Profile);
+
+            if (const auto tabColor{ JsonUtils::GetValueForKey<std::optional<til::color>>(json, TabColorKey) })
```
I think if you make TabColor an `Windows::Foundation::IReference<Windows::UI::Color>`, like `Profile::Foreground`, you should be able to just do 
```c++
    JsonUtils::GetValueForKey(json, TabColorKey, _TabColor );
```

Don-Vito replied:
@zadjii-msft - actually I started with TabColor being a Windows::UI::Color, but then I saw that SetTabColor command uses til:color, thus I switched to it in the sake of code consistency. I suggest to either leave it as is or I can fix the SetTabColor command to use Windows::UI::Color. WDYT?

zadjii-msft replied:
Ah heck, that should probably be a `Windows::UI::Color` too. Sorry to bounce you around like this!

Don-Vito replied:
@zadjii-msft - NP. I am fixing this!

zadjii-msft commented on src/cascadia/TerminalApp/Resources/en-US/Resources.resw:
```diff
@@ -285,6 +285,9 @@
     <comment>{Locked="\"title\""}</comment>
   </data>
+  <data name="CmdTabColorArgDesc" xml:space="preserve">
+    <value>Open the terminal with the tab color</value>
```
nit:
```suggestion
    <value>Open the terminal with the given tab color, in #rrggbb format</value>
```

zadjii-msft reviewed (changes requested):
These are kinda just nits, but I do think the `Windows::UI::Color` thing would be slightly cleaner.

Otherwise, great job here!

Don-Vito commented:
@zadjii-msft - fixed, please re-review :blush:

zadjii-msft reviewed (approved):
Yea, this is the stuff. Great work!

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionArgs.h:
```diff
@@ -513,10 +518,7 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         {
             // LOAD BEARING: Not using make_self here _will_ break you in the future!
             auto args = winrt::make_self<SetTabColorArgs>();
-            if (const auto temp{ JsonUtils::GetValueForKey<std::optional<til::color>>(json, ColorKey) })
-            {
-                args->_TabColor = static_cast<uint32_t>(*temp);
-            }
+            JsonUtils::GetValueForKey(json, ColorKey, args->_TabColor);
```
Thank you for this :)

carlos-zamora commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -645,7 +651,7 @@ namespace winrt::TerminalApp::implementation
     //      currently displayed, it will be shown.
     // Arguments:
     // - settings: the TerminalSettings object to use to create the TerminalControl with.
```
nit: update the comments up here for this function to say that you're returning a tab now

carlos-zamora reviewed (changes requested):
This is awesome! Just link the PR for the updated docs and I'll approve. Thanks!

DHowett commented on src/cascadia/TerminalApp/Resources/en-US/Resources.resw:
```diff
@@ -295,6 +295,9 @@
     <comment>{Locked="\"title\""}</comment>
   </data>
+  <data name="CmdTabColorArgDesc" xml:space="preserve">
+    <value>Open the terminal with the tab color, in #rrggbb format</value>
```
```suggestion
    <value>Open this tab with the specified tab color, in #rrggbb format</value>
```

(since this is an argument to nt/new-tab you can refer to an implicit tab. Panes technically.. won't quite work like this (!))

DHowett reviewed (changes requested):
hmm. what happens when you specify a tab color on split-pane? It looks like it becomes valid, but does nothing

Don-Vito commented:
> 
> 
> hmm. what happens when you specify a tab color on split-pane? It looks like it becomes valid, but does nothing

@DHowett - thanks! I am not sure what should be the behavior. I even started discussing this with @zadjii-msft in the ticket itself, but I believe we never finished the discussion.

Don-Vito commented:
@zadjii-msft, @DHowett - I have created an [alternative implementation](https://github.com/Don-Vito/terminal/pull/4) where I add tabColor only to NewTabArgs. Please let me know if you prefer that variant, or you believe we do need to add tabColor for split-pane. In the latter case I need your help with understanding the expected behavior.

Don-Vito commented:
> 
> 
> This is awesome! Just link the PR for the updated docs and I'll approve. Thanks!

@carlos-zamora - thanks for the review! I will update the docs, once we decide what to do with the split-pane :blush:

Don-Vito commented:
> 
> 
> @zadjii-msft, @DHowett - I have created an [alternative implementation](https://github.com/Don-Vito/terminal/pull/4) where I add tabColor only to NewTabArgs. Please let me know if you prefer that variant, or you believe we do need to add tabColor for split-pane. In the latter case I need your help with understanding the expected behavior.

Bump - so I can advance on week-end

aj1988aj commented on src/cascadia/LocalTests_SettingsModel/KeyBindingsTests.cpp:
```diff
@@ -408,7 +408,7 @@ namespace SettingsModelLocalTests
             VERIFY_IS_NOT_NULL(realArgs.TabColor());
             // Remember that COLORREFs are actually BBGGRR order, while the string is in #RRGGBB order
-            VERIFY_ARE_EQUAL(static_cast<uint32_t>(til::color(0x563412)), realArgs.TabColor().Value());
+            VERIFY_ARE_EQUAL(til::color(0x563412), til::color(realArgs.TabColor().Value()));
```
```suggestion
            VERIFY_ARE_EQUAL(til::color(0x563412), til::color(realArgs.TabColor().Value()));
```

Don-Vito replied:
@aj1988aj - I am not sure what is the suggested diff. Spaces look OK.

DHowett commented:
So it's interesting. A profile can set a tabColor, which means that panes _technically have one_. I'll need @zadjii-msft to weigh in.

It might boil down to "it should be like a title -- the focused pane wins -- we just didn't do it like that"

Don-Vito commented:
> So it's interesting. A profile can set a tabColor, which means that panes _technically have one_. I'll need @zadjii-msft to weigh in.
> 
> It might boil down to "it should be like a title -- the focused pane wins -- we just didn't do it like that"

@DHowett - cool. I will try to implement this approach as well and then we can choose :blush:

Don-Vito commented:
@DHowett , @zadjii-msft - changed the implementation to set the color on the pane level. 
However, I could not use existing hierarchy:
* runtimeTabColor - is not good enough as it is on the tab level
* Terminal::TabColor - is not good enough, as it gets overridden upon settings reload

Hence I added a StartingTitle to TermControl, I was not sure if I should add it to TermControl or Terminal itself (I will probably move it to Terminal in a separate PR, and merge to here if required)

Don-Vito commented:
I decided to move everything to Terminal, we'll see if it was a good decision 😄

Don-Vito commented:
@zadjii-msft - I added the starting tab color to be a property of the Terminal. Consequently, when a new tab is created the color is set on the first pane. But if we split the tab the color is not set on the second pane. We could say that the color will be applied for all panes of the tab, but this gets complicated, as the color can be also specified on the second pane as well. Which will require us to resolve what color is preferable. This is possible, but I am not sure if we want to get there product-wise. WDYT?

Don-Vito commented:
> 
> 
> This is awesome! Just link the PR for the updated docs and I'll approve. Thanks!

Updated the docs. Thanks!

zadjii-msft commented:
Okay, sorry for the delay reviewing this. You're churning out so many dang PRs, it's hard to keep up 😄 

I've played with the branch as it is this morning, and I like how it feels. And it feels like what I originally described in #8075. Plus, being resilient to a settings reload is a clever solution. 

I'll sanity check the code now, but it definitely _feels_ 👌

zadjii-msft reviewed (approved):
Oh yea I already reviewed 90% of this. This looks good to me!

carlos-zamora commented on src/cascadia/TerminalCore/Terminal.cpp:
```diff
@@ -1159,9 +1165,12 @@ void Terminal::ClearPatternTree() noexcept
 
+// Method Description:
+// - Returns the tab color
+// If the starting color exits it's value is preferred
```
```suggestion
// If the starting color exists, it's value is preferred
```

Don-Vito replied:
thanks!

carlos-zamora commented on src/cascadia/TerminalApp/TerminalSettings.h:
```diff
@@ -69,6 +69,7 @@ namespace winrt::TerminalApp::implementation
         GETSET_PROPERTY(bool, CopyOnSelect, false);
 
         GETSET_PROPERTY(Windows::Foundation::IReference<uint32_t>, TabColor, nullptr);
+        GETSET_PROPERTY(Windows::Foundation::IReference<uint32_t>, StartingTabColor, nullptr);
```
@DHowett or @zadjii-msft:
Should we have a comment here or somewhere describing the difference between these two properties? I understand that we _need_ `StartingTabColor`, but I have a feeling it'll be confusing to have both down the road.

Don-Vito replied:
@carlos-zamora - I totally agree with you that it is both confusing and error-prone. Instead I would like to:
1. Introduce "ITerminalCreationParams" and move this property there.
2. Pass them upon construction

However this will require changing the Terminal interface, that I felt somewhat above my paygrade at this  stage.

carlos-zamora reviewed (approved):
I guess I'm ok with this. I like the feature. I don't like the `TabColor` vs `StartingTabColor` thing, but I'll just have to live with it since it seems to be the best approach tbh.

DHowett reviewed (approved):
This is great. We'll need to get the TerminalSettings back into check because sometimes we use it to communicate app-only state, but I'm happy with this for now.
